Approve console cmdlets tests

iSazonov opened the pull request:
Main improvements refer to tests of the Write-Host cmdlet.
Original tests:
1. Slow because run external processes
2. Don't test colors and -NoNewLine in fact.

Done:
1. The original tests is preserved (deleted one as redundant) but marked
by 'Slow' tag. They is preserved because they actually check the output
on the work, not a test console.
2. Add negative color tests. (Code cover grow!)
3. Add tests based on TestHostCS. This test host has been refined so we
can see colors and a new line in output.
4. Add minor fixes for test modules loads.

Also I add support for Information stream. I originally planned to use
it but not actually used. However, I have left this as a useful addition
for future tests.
I wonder that a Write-Host console output is duplicated in Information
Stream - Is it by design? I left a debug print on this matter in the
test code.

iSazonov committed:
Approve console cmdlets tests

Main improvements refer to tests of the Write-Host cmdlet.
Original tests:
1. Slow because run external processes
2. Don't test colors and -NoNewLine in fact.

1. The original tests is preserved (deleted one as redundant) but marked
by 'Slow' tag. They is preserved because they actually check the output
on the work, not a test console.
2. Add negative color tests. (Code cover grow!)
3. Add tests based on TestHostCS. This test host has been refined so we
can see colors and a new line in output.
4. Add minor fixes for test modules loads.

Also I add support for Information stream. I originally planned to use
it but not actually used. However, I have left this as a useful addition
for future tests.
I wonder that a Write-Host console output is duplicated in Information
Stream - Is it by design? I left a debug print on this matter in the
test code.

lzybkr commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -1,5 +1,14 @@
-﻿Import-Module $PSScriptRoot\..\..\Common\Test.Helpers.psm1
-Import-Module $PSScriptRoot\..\..\Common\TestHostCS.psm1
+﻿if ( ! (get-module -ea silentlycontinue Test.Helpers ))
```
Why?  `Import-Module` does this already, right?

iSazonov replied:
"Don't trust code patterns... Don't trust code patterns... Don't trust code patterns..." :-)
I am sorry but I cleaned up all files with bad pattern.

lzybkr commented on test/powershell/Common/TestHostCS.psm1:
```diff
@@ -290,7 +304,7 @@ function New-TestHost
 
     if ( ! ("TestHost.TestHost" -as "type" )) {
-       $t = add-Type -pass $definition -ref $references
+       $t = add-Type -pass $definition -ref $references -WarningAction SilentlyContinue 
```
What warnings are being suppressed?  I assume it's a C# warning - maybe disable with a [pragma](https://msdn.microsoft.com/en-us/library/441722ys.aspx) instead so we don't ignore future warnings that maybe shouldn't be ignored.

iSazonov replied:
I check CI logs - no warnings. I see the warnings only locally. 
So revert the change.

lzybkr commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,93 @@
+        $ps.Runspace = $rs
+
+        $testHostCSData = @(
+            @{ Name = 'defaults';Command = "Write-Host a,b,c"; returnCount = 1; returnValue = @("White:Black:a b c:NewLine") }
```
I think this would be easier to read and review if each key/value pair was on a line by itself.

iSazonov replied:
Fixed.

lzybkr commented on test/powershell/Common/TestHostCS.psm1:
```diff
@@ -191,7 +196,7 @@ namespace TestHost
         public override void WriteLine(string value)
         {
-            Streams.ConsoleOutput.Add(value);
+            Streams.ConsoleOutput.Add("::"+value+":NewLine");
```
Why not use `` `n``, or I guess since this is actually C#, `\n`? Then you don't need `:NoNewline`.

iSazonov replied:
Idea is that we shouldn't mix "user" chars and "code" chars. If we would use "`n" and `value` is "test`n" then we will not be able to determine whether this character from the original string or generated by code (WriteLine).
We could use "test:`n" or "test:\n" but if we look at the text of tests, it is easier to read `NewLine` and `NoNewLine`, especially since this is key information.
We only need to avoid using ':' in `value`. I believe that this is not a problem for test environment.

daxian-dbw replied:
Is the preceding `::` necessary?

iSazonov replied:
This means that colors are not specified. We should use one template for all outputs. In the future someone can use "-Split" to analyze results regardless of which Write/WriteLine method is used.

daxian-dbw replied:
If color is specified, I would use `color:output` as the expected string to test the result, so why `::` is needed?

iSazonov replied:
Here colors is _not_ specified and it is _Empty_. I believe both output methods (with colors and w/o colors) should return strings in the same format. I would make "Empty:Empty:teststring:NewLine" but it is seems unneeded complicity.

daxian-dbw replied:
I see your point now. It sounds over complex and useful at the same time 😄 
I'm OK to have this change in TestHostCS.psm1, but could you please put more comments there for the `Write/WriteLine` APIs so that people can a bit easier to understand the behavior?

iSazonov replied:
Done.

iSazonov committed:
Fix after code review

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Core/Out-Host.Tests.ps1:
```diff
@@ -27,8 +21,9 @@ Describe "Out-Host Tests" -tag CI {
     }
     It "Out-Host writes to host output" {
         $stringToWrite = "thing to write"
+        $stringExpected = "::$($stringToWrite):NewLine"
```
Minor: the sub expression `$()` is not needed here.

iSazonov replied:
Without $() it is not worked 😕

daxian-dbw replied:
Ah, I see, the preceding colons. Again, are the preceding colons necessary?

daxian-dbw replied:
No, it's the colon at `:NewLine`. Maybe use another special character as the delimiter?

iSazonov replied:
This is debug output - we can use any perfectly visible symbol.
The only restriction is that not to use it in the source debug string (<value>).
This character ':' is rarely used, so I suppose it is appropriate.
Or I missed something?

daxian-dbw replied:
It's just less intuitive to construct the expected string if you are using a variable. But currently there is only one instance where `$()` is required. Let's keep it for now, and we can always change to something else if there are any complaints.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,91 @@
-﻿Describe "Write-Host DRT Unit Tests" -Tags "CI" {
+﻿Describe "Write-Host DRT Unit Tests" -Tags "Slow","Feature" {
```
> "Write-Host DRT Unit Tests"

Can you change this description to be more specific? Other tests here seem also DRT unit tests.

iSazonov replied:
Fixed.

Maybe remove `Feature` tag?

daxian-dbw replied:
I found this:
> Microsoft.PowerShell.Management\Start-Process.Tests.ps1:1:Describe "Start-Process" -Tags @("CI","SLOW") {

So let's keep both slow and feature here.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,91 @@
+        [array]$result = & $powershell -noprofile $Command
+        $result.Count | Should Be $returnCount
+        $result[0] | Should Be $returnValue[0]
+        $result[1] | Should Be $returnValue[1]
```
It's unlikely new test data will be aded to `$testData`, but I think it's better to use the same `foreach ($i in 0..($returnCount - 1))` pattern (or something similar).

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,91 @@
 
-    It "write-Host works with '<Name>' switch" -TestCases $testData -Pending:$IsOSX {
-        param($Command, $returnValue)
+        # I wonder that a console output is duplicated in Information Stream - Is it by design?
```
👍 Thanks!

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,91 @@
+        # I wonder that a console output is duplicated in Information Stream - Is it by design?
+        Write-Host $th.ui.Streams.Information.Count
+        Write-Host $th.ui.Streams.Information[0]
+        Write-Host $th.ui.Streams.Information[1]
```
Currently we haven't test for the "by design" feature so we should convert the lines to real tests.

iSazonov replied:
Done.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Host.Tests.ps1:
```diff
@@ -1,18 +1,91 @@
-        & $powershell -noprofile $Command | Should Be $returnValue
+        $result.Count | Should Be $returnCount
+        $result[0] | Should Be $returnValue[0]
+        $result[1] | Should Be $returnValue[1]
```
What scenarios do you have in mind? It seems these tests verify everything we need and look simple and clear.

daxian-dbw replied:
Thanks for making this change!
Yes, tests for Write-Host is thorough enough, but it's better to set a good pattern for other people to copy in their tests.

iSazonov replied:
I totally agree!

iSazonov committed:
Corrections after code review

Suppress import-module warnings
Rename Describes
Add "-Object" test
Add Stream.Information tests with TestHostCS

iSazonov commented:
@PowerShellTeam We have a problem with CIs:
> Failed to register repository 'mygetpsmodule'

daxian-dbw commented:
@iSazonov weird, I do see two of your PRs ran into this break, but no others. The PR #3139 was opened an hour ago and the builds were successful. Is this an intermittent problem? I restarted the Linux CI and will keep an eye on it.

daxian-dbw commented:
I re-ran the Linux Ci and this time there is no build break. It's possible that the myget feed we used for installing `PowerShellGet` and `PackageManagement` was having a hiccup 😦

iSazonov committed:
Add checks for Streams.Information and add comments

daxian-dbw reviewed (approved):
LGTM

iSazonov commented:
@daxian-dbw Thanks for code review!
